### PR TITLE
En 113

### PR DESCRIPTION
**Pull request by FedericoA15**
# Description :pencil:

Endpoint was created to synchronize existing birthdays in Google Calendar and when an employee is created it saves the birthday date

## Link to ticket :link:

https://entropyteam.atlassian.net/browse/EN-113

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested? :microscope:

image

### CODE REVIEW

**Review by jauday — APPROVED** [review]

File: pom.xml [hunk 1]
```diff
@@ -1,164 +1,174 @@
...
+	  <version>1.1.15-SNAPSHOT</version>
+	  <packaging>jar</packaging>
+    <name>entropay-employees</name>
+    <description>Entropay Employees API</description>
+    <properties>
+        <apache-commons.version>3.0</apache-commons.version>
+        <apache-commons-collections.version>4.0</apache-commons-collections.version>
+        <flywaydb.version>8.5.10</flywaydb.version>
+        <google-auth.version>1.20.0</google-auth.version>
+        <google-oauth-client-jetty>1.34.1</google-oauth-client-jetty>
```
Comment (agustinb4) [inline pom.xml:22]: not used

File: src/main/java/com/entropyteam/entropay/employees/controllers/EmployeeController.java [hunk 2]
```diff
@@ -1,26 +1,48 @@
...
 
+    private final EmployeeService employeeService;
 
     @Autowired
     public EmployeeController(EmployeeService employeeService) {
         super(employeeService);
+        this.employeeService = employeeService;
     }
 
+    @GetMapping("/birthday")
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/controllers/EmployeeController.java:36]: minor, can we change the endpoint to /sync-birthday. Also the method name could be syncEmployeeBirthday

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        Calendar service =
+                new Calendar.Builder(HTTP_TRANSPORT, JSON_FACTORY, new HttpCredentialsAdapter(getCredentialsServiceAccount()))
+                        .setApplicationName("GoogleCalender")
+                        .build();
+        CalendarListEntry calendarListEntry = new CalendarListEntry();
+        calendarListEntry.setId("<Calender Id>");
+        CalendarListEntry updatedCalendarList = service.calendarList().insert(calendarListEntry).execute();
+        LOGGER.info("Calendar Summary: {}", updatedCalendarList.getSummary());
+    }
+}
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java:138]: remove

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                    if (start == null) {
+                        start = event.getStart().getDate();
+                    }
+                    System.out.printf("%s (%s)\n", event.getSummary(), start);
+                }
+            }
+        } catch (IOException | GeneralSecurityException e) {
+            e.printStackTrace();
+        }
+    }
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java:87]: remove

File: src/test/java/com/entropyteam/entropay/employees/services/GoogleServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package com.entropyteam.entropay.employees.services;
```
Comment (agustinb4) [inline src/test/java/com/entropyteam/entropay/employees/services/GoogleServiceTest.java:1]: remove test

File: src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java [hunk 6]
```diff
@@ -213,4 +220,15 @@ private boolean shouldDeactivateEmployee(UUID employeeId, Employee entityToUpdat
         Employee existingEmployee = getRepository().getById(employeeId);
         return existingEmployee.isActive() && !entityToUpdate.isActive();
     }
+
+    public void createGoogleCalendarEventsForBirthdays() throws IOException {
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java:224]: will this method create duplicates?

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            DateTime startDateTime = new DateTime(startDateStr);
+            DateTime endDateTime = new DateTime(endDateStr);
+
+            EventDateTime startEventDateTime = new EventDateTime().setDate(startDateTime);
+            EventDateTime endEventDateTime = new EventDateTime().setDate(endDateTime);
+
+            event.setStart(startEventDateTime);
+            event.setEnd(endEventDateTime);
+
+            String calendarId = "<Calender Id>";
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java:114]: this should be a env var

File: src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java [hunk 8]
```diff
@@ -109,6 +113,9 @@ public EmployeeDto create(EmployeeDto employeeDto) {
         Employee entityToCreate = toEntity(employeeDto);
         Employee savedEntity = getRepository().save(entityToCreate);
         paymentInformationService.createPaymentsInformation(savedEntity.getPaymentsInformation(), savedEntity);
+        LocalDate birthDate = employeeDto.birthDate();
+        LocalDate modifiedBirthDate = birthDate.withYear(LocalDate.now().getYear());
+        googleService.createGoogleCalendarEvent("BirthDay " + employeeDto.firstName() + " " + employeeDto.lastName(), modifiedBirthDate, modifiedBirthDate);
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java:118]: minor, lets extract this lines in a method that receives a list of employees, create the event name then call google service

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            event = service.events().insert(calendarId, event).execute();
+
+            System.out.println("Event created: " + event.getHtmlLink());
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java:117]: use Logger

**Review by agustinb4 — CHANGES_REQUESTED** [review]

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+                    if (start == null) {
+                        start = event.getStart().getDate();
+                    }
+                    System.out.printf("%s (%s)\n", event.getSummary(), start);
+                }
+            }
+        } catch (IOException | GeneralSecurityException e) {
+            e.printStackTrace();
+        }
+    }
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java:78]: remove

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
+            String startDateStr = dateFormat.format(startDateUtil);
+            String endDateStr = dateFormat.format(endDateUtil);
+
+            DateTime startDateTime = new DateTime(startDateStr);
+            DateTime endDateTime = new DateTime(endDateStr);
+
+            EventDateTime startEventDateTime = new EventDateTime().setDate(startDateTime);
+            EventDateTime endEventDateTime = new EventDateTime().setDate(endDateTime);
+
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java:102]: avoid multiple convertions

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        Calendar service = new Calendar.Builder(HTTP_TRANSPORT, JSON_FACTORY,
+                new HttpCredentialsAdapter(getCredentialsServiceAccount()))
+                .setApplicationName("GoogleCalender")
+                .build();
+        CalendarListEntry calendarListEntry = new CalendarListEntry();
+        calendarListEntry.setId("<Calender Id>");
+        CalendarListEntry updatedCalendarList = service.calendarList().insert(calendarListEntry).execute();
+        LOGGER.info("Calendar Summary: {}", updatedCalendarList.getSummary());
+    }
+}
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java:129]: remove

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+
+            event.setStart(startEventDateTime);
+            event.setEnd(endEventDateTime);
+
+            String calendarId = "<Calender Id>";
+            event = service.events().insert(calendarId, event).execute();
+
+            System.out.println("Event created: " + event.getHtmlLink());
+        } catch (IOException | GeneralSecurityException e) {
+            e.printStackTrace();
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java:111]: fix with LOGGER.error(e.getMessage());

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleCredentialsProperties.java [hunk 14]
```diff
@@ -0,0 +1,14 @@
+package com.entropyteam.entropay.employees.services;
+
+import lombok.Data;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties("google-credentials")
+@Data
+public class GoogleCredentialsProperties {
+    private String idClient;
+    private String secretClient;
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleCredentialsProperties.java:10]: remove

File: src/main/resources/application.properties [hunk 15]
```diff
@@ -21,4 +21,10 @@ spring.security.oauth2.resourceserver.jwt.issuer-uri=''
 # Aws credentials
 aws-credentials.access-key-id=''
 aws-credentials.secret-access-key=''
-aws-credentials.bucket-name=''
\ No newline at end of file
+aws-credentials.bucket-name=''
+
+# Google credentials
+google-credentials.id-client=''
+google-credentials.secret-client=''
```
Comment (agustinb4) [inline src/main/resources/application.properties:28]: remove

File: src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java [hunk 16]
```diff
@@ -216,7 +225,18 @@ private boolean shouldDeactivateEmployee(UUID employeeId, Employee entityToUpdat
         return existingEmployee.isActive() && !entityToUpdate.isActive();
     }
 
-    public String getEmployeesTimeSinceStart(Contract firstContract, Contract latestContract){
+    public CalendarEventDto formatEventData(UUID employeeId, LocalDate birthDate, String firstName, String lastName) {
+        int currentYear = LocalDate.now().getYear();
+        LocalDate startDate = birthDate.withYear(currentYear);
+        LocalDate endDate = startDate.plusDays(1);
+        String eventId = currentYear + "-" + employeeId.toString();
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/EmployeeService.java:232]: You are adding a "-" here, then replacing it for a blank space in the google service. Lets change to:
`String eventId = currentYear + employeeId.toString();`

File: src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+public class GoogleService {
+    private final GoogleCredentialsProperties googleCredentialsProperties;
+    private static final Logger LOGGER = LogManager.getLogger();
+    private static final JsonFactory JSON_FACTORY = GsonFactory.getDefaultInstance();
+    private static final List<String> SCOPES = Collections.singletonList(CalendarScopes.CALENDAR);
+
+    @Autowired
+    public GoogleService(GoogleCredentialsProperties googleCredentialsProperties) {
+        this.googleCredentialsProperties = googleCredentialsProperties;
+    }
```
Comment (agustinb4) [inline src/main/java/com/entropyteam/entropay/employees/services/GoogleService.java:40]: you can change to:
   ```
 
    private static final Logger LOGGER = LogManager.getLogger();
    private static final JsonFactory JSON_FACTORY = GsonFactory.getDefaultInstance();
    private static final List<String> SCOPES = Collections.singletonList(CalendarScopes.CALENDAR);

    @Autowired
    private final GoogleCredentialsProperties googleCredentialsProperties;

```
removing the contructor

**Review by agustinb4 — APPROVED** [review]
Good Job man! some minors comments

**Pull request merged**